trunk/source/Library-ModelCheckerUtils/src/de/uni_freiburg/informatik/ultimate/lib/modelcheckerutils/cfg/ProcedureMultiplier.java

```diff
@@ -86,182 +84,173 @@
+				new HashRelation<>(cfgSmtToolkit.getModifiableGlobalsTable().getProcToGlobals());
 
-		final Map<ILocalProgramVar, ILocalProgramVar> newVar2OldVar = new HashMap<>();
+		// Add variables
 		final Map<String, Map<ILocalProgramVar, ILocalProgramVar>> oldVar2newVar = new HashMap<>();
 		final Map<Term, Term> variableBacktranslationMapping = new HashMap<>();
-		managedScript.lock(this);
+		managedScript.lock(lockOwner);
 		for (final String proc : copyDirectives.getDomain()) {
-			assert icfg.getCfgSmtToolkit().getProcedures().contains(proc) : "procedure " + proc + " missing";
+			assert cfgSmtToolkit.getProcedures().contains(proc) : "procedure " + proc + " missing";
 			for (final String copyIdentifier : copyDirectives.getImage(proc)) {
 				icfg.addProcedure(copyIdentifier);
 				final Map<ILocalProgramVar, ILocalProgramVar> procOldVar2NewVar = new HashMap<>();
 				procedures.add(copyIdentifier);
 				final List<ILocalProgramVar> inParamsOfCopy = new ArrayList<>();
 				for (final ILocalProgramVar inParam : inParams.get(proc)) {
 					final ILocalProgramVar inParamCopy = constructVariableCopy(inParam, copyIdentifier, managedScript,
-							procOldVar2NewVar, variableBacktranslationMapping, symbolTable);
+							procOldVar2NewVar, variableBacktranslationMapping, symbolTable, lockOwner);
 					inParamsOfCopy.add(inParamCopy);
-					newVar2OldVar.put(inParamCopy, inParam);
 				}
 				inParams.put(copyIdentifier, inParamsOfCopy);
 				final List<ILocalProgramVar> outParamsOfCopy = new ArrayList<>();
 				for (final ILocalProgramVar outParam : outParams.get(proc)) {
 					final ILocalProgramVar outParamCopy = constructVariableCopy(outParam, copyIdentifier, managedScript,
-							procOldVar2NewVar, variableBacktranslationMapping, symbolTable);
+							procOldVar2NewVar, variableBacktranslationMapping, symbolTable, lockOwner);
 					outParamsOfCopy.add(outParamCopy);
-					newVar2OldVar.put(outParamCopy, outParam);
 				}
 				outParams.put(copyIdentifier, outParamsOfCopy);
-				for (final ILocalProgramVar localVar : icfg.getCfgSmtToolkit().getSymbolTable().getLocals(proc)) {
+				for (final ILocalProgramVar localVar : cfgSmtToolkit.getSymbolTable().getLocals(proc)) {
 					if (!procOldVar2NewVar.containsKey(localVar)) {
 						constructVariableCopy(localVar, copyIdentifier, managedScript, procOldVar2NewVar,
-								variableBacktranslationMapping, symbolTable);
+								variableBacktranslationMapping, symbolTable, lockOwner);
 					}
 				}
-				final List<IProgramNonOldVar> modifiableGlobals = new ArrayList<>();
-				modifiableGlobals.addAll(proc2globals.getImage(proc));
-				for (final IProgramNonOldVar modifiableGlobal : modifiableGlobals) {
-					proc2globals.addPair(copyIdentifier, modifiableGlobal);
-				}
 				oldVar2newVar.put(copyIdentifier, procOldVar2NewVar);
+				proc2globals.getImage(proc).forEach(g -> proc2globals.addPair(copyIdentifier, g));

```

Comment:
Does this work with `ConcurrentModificationException`?

Reply:
It should work, since `HashRelation::getImage` returns a `Collections.unmodifiableSet`

Reply:
Yeah I didn't realize you're not iterating over the relation itself here